src/openms/source/CHEMISTRY/Tagger.cpp

```diff
@@ -0,0 +1,201 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Timo Sachsenberg $
+// $Authors: Timo Sachsenberg $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/CHEMISTRY/Tagger.h>
+#include <OpenMS/CHEMISTRY/ResidueDB.h>
+#include <OpenMS/CHEMISTRY/Residue.h>
+#include <OpenMS/CHEMISTRY/ModificationsDB.h>
+#include <OpenMS/MATH/MISC/MathFunctions.h>
+
+namespace OpenMS
+{
+  std::string Tagger::getAAByMass_(float m) const
+  {
+     // fast check for border cases
+     if (m < min_gap_ || m > max_gap_) return " ";
+     const float delta = Math::ppmToMass(ppm_, m);
+     auto left = mass2aa.lower_bound(m - delta);
+     //if (left == mass2aa.end()) return " "; // cannot happen, since we checked boundaries above
+
+     // Check for multiple charges, prioritize lower charges (returns with the lowest charge with a match).
+     // This assumes both fragments had the same charge,
+     // but ignores whether ions in the longer sequence all have the same charge.
+     // This could lead to inconsistencies where an ion is considered to have different charges when compared to its left and right neighbors in the sequence (TODO?).
+     // This looks at the difference between two charged ions, so the proton mass is the same on both ions and can be ignored here.
+     for (size_t charge = min_charge_; charge <= max_charge_; ++charge)
+     {
+       double charged_m = m / charge;
+       if (fabs(left->first - charged_m) < delta) return left->second;
+     }
+     return " ";
+  }
+
+  void Tagger::getTag_(StringList & tag, const std::vector<float>& mzs, const size_t i, std::set<std::string>& tags) const
+  {
+    const size_t N = mzs.size();
+    size_t j = i + 1;
+    // recurse for all peaks in distance < max_gap
+    while (j < N)
+    {
+      if (tag.size() == max_tag_length_) { return; } // maximum tag size reached? - continue with next parent
+
+      const float gap = mzs[j] - mzs[i];
+      //cout << i << "\t" << j << "\t" << mzs[i] << "\t" << mzs[j] << "\t" << gap << endl;
+      if (gap > max_gap_) { return; } // already too far away - continue with next parent
+      const std::string aa = getAAByMass_(gap);
+      if (aa == " ") { ++j; continue; } // can't extend tag
+      tag.push_back(aa);
+      getTag_(tag, mzs, j, tags);
+      // each tag is build as a StringList, so that push_back(), pop_back() and size() work as intended even with modified residues, e.g. "M(Oxidation)"
+      // and then the tag is reported as a string by collapsing it here using concatenate()
+      if (tag.size() >= min_tag_length_) tags.insert(ListUtils::concatenate(tag));
+
+      // if aa is "L", then also add "I" as an alternative residue and extend the tag again
```

Comment:
why not use "J" in tags and convert all occurences of I/L in proteins to J before comparison?